### PR TITLE
Compartmentalizing

### PR DESCRIPTION
**Pull request by ghazalemir**
# Description of the problem

Compartment modelling (CM) provides a simpler and less computationally demanding approach compared to CFD, in which flows are defined between a network of ideally mixed zones called compartments. With compartment models the modeler must make decisions regarding the number of compartments, the volumes of the compartments, the connections between the compartments, and the flow rates between them. (Ref: Automated Compartment Model Development Based on Data from Flow-Following Sensor Devices)

# Description of the solution

This code identifies the network of compartments, cornering two groups of target variables,
It first divide the domain to compartments based on Electric Field data,
Next, it finds the overlaid map of Electric field and velocity

# How Has This Been Tested?
There is an example (lethe/examples/rpt/compartmentalization) containing the EF and velocity data at the cell centers for the given geometry and mesh

# Future changes
-Adding the parameter file
-Adding the volume of compartments calculation
-Adding the flow rate between compartments calculation

# Comments

- This is the first commit of mine and I not familiar with the structure of how a code should look like apart from its performance
- Please help me to learn and improve :)

### CODE REVIEW

File: include/rpt/compartmentalization.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+private:
+  /**
+   * @brief Read the physical property (in this problem, the electric field)
+   * and write it into a vector of a vector.
+   */
+  auto
+  read_physical_property() -> std::vector<std::vector<double>>;
+
+  /**
+   * @brief Read the velocity magnitude and write it into a vector of a vector.
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:66]: Does it create a vector of a vector or a map?

File: include/rpt/compartmentalization.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+   * @brief Generate subdivided_cylinder mesh
+   */
+  void
+  generate_cylindrical_grid();
+
+  /**
+   * @brief 1) Sorts the cell based on their
+   *  values of the physical properties
+   *  (average physical values, in particular electric field)
+   *  2) agglomerate cell within a tolerance and then 3) break the
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:81]:
```suggestion
   *  2) agglomerates cell within tolerance and then 3) break the
```

File: include/rpt/compartmentalization.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  /**
+   * @brief Generate subdivided_cylinder mesh
+   */
+  void
+  generate_cylindrical_grid();
+
+  /**
+   * @brief 1) Sorts the cell based on their
+   *  values of the physical properties
+   *  (average physical values, in particular electric field)
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:80]:
```suggestion
   *  (average physical values, in particular, electric field)
```

File: include/rpt/compartmentalization.h (hunk 2 above)
Comment (voferreira) [inline include/rpt/compartmentalization.h:81]:
```suggestion
   *  2) agglomerate cell within tolerance and then 3) break the
```

File: include/rpt/compartmentalization.h [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+   *  2) agglomerate cell within a tolerance and then 3) break the
+   * clusters that they are not connected but they are in one group
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:82]:
```suggestion
   * clusters that are not connected but are in one group
```

File: include/rpt/compartmentalization.h [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+   * because of their tolerance.
+   */
+  auto
+  sort_agglomeration_deagglomeration_emw() -> std::map<int,std::vector<typename Triangulation<dim>::active_cell_iterator>>;
+
+  /**
+   * @brief this function get the first set of compartments based on one
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:89]:
```suggestion
   * @brief this function gets the first set of compartments based on one
```

File: include/rpt/compartmentalization.h [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+   * of the physical values (for example, electric field) from
+   * "sort_agglomeration_deagglomeration_emw()", for each compartment
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:91]:
```suggestion
   * "sort_agglomeration_deagglomeration_emw()" for each compartment
```

File: include/rpt/compartmentalization.h [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+   * this function treat each compartment as a new problem to be compartmentalized
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:92]:
```suggestion
   * this function treats each compartment as a new problem to be compartmentalized
```

File: include/rpt/compartmentalization.h [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+   * and based on the second physical property (velocity)
+   * it 1) sorts the cell, 2) agglomerate within the tolerance, 3)deagglomerate
```
Comment (voferreira) [inline include/rpt/compartmentalization.h:94]:
```suggestion
   * it 1) sorts the cell, 2) agglomerate within the tolerance, 3) deagglomerate
```

File: source/rpt/compartmentalization.cc [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+    }
+  //output the center of cell to feed the Pyvista and extract the physical
+  //cfd or other simulation
+  std::string filename ="cell_centers";
+  std::ofstream myfile;
+  std::string   sep = " ";
+  myfile.open(filename);
+  for(auto & point : points){
+      for(int j=0;j<3;j++){
+          myfile << point[j]<< sep;
```
Comment (voferreira) [inline source/rpt/compartmentalization.cc:56]:
```suggestion
          myfile << point[j] << sep;
```

File: source/rpt/compartmentalization.cc [hunk 10]
```diff
@@ -0,0 +1,509 @@
+#include <deal.II/grid/cell_id.h>
+#include <deal.II/numerics/data_component_interpretation.h>
```
Comment (voferreira) [inline source/rpt/compartmentalization.cc:2]: The same way as the .h, this could be reordered to

**Review by voferreira — COMMENTED** [review]
Another checkpoint. I will finish this today :)

File: source/rpt/compartmentalization.cc [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+
+  write_vtu_and_pvd<dim>(
+    grid_pvdhandler,
+    data_out,
+    folder,
+    "emw",
+    time,
+    step_number,
+    1, mpi_communicator);
+}
```
Comment (voferreira) [inline source/rpt/compartmentalization.cc:462]:
```suggestion
}

```

File: source/rpt/compartmentalization.cc [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+          compartments_emw[d->active_cell_index()]=compartment_final_id;
+        }
+      compartment_final_id=compartment_final_id+1;
+    }
+  write_file_compartments_first_field(
+    compartments_emw,
+    1.0,
+    1);
+  return final_set;
+}
```
Comment (voferreira) [inline source/rpt/compartmentalization.cc:273]:
```suggestion
}

```

**Review by voferreira — COMMENTED** [review]
This is really awesome, friend! My comments are mostly related to organizing the code a bit. I would have broken down the function sort_agglomeration_deagglomeration_emw into functions to make it a bit more readable. If you want, I can do another pass after the others' review.

**Comment by voferreira** [comment]
@ghazalemir Whenever you will address the comments, please let me help you. There is a way to do this more efficiently on GitHub :)

**Review by Luckabarbeau — COMMENTED** [review]
first batch

File: source/rpt/compartmentalization.cc [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+  int number_of_compartments =1;
+  //this "k" iterator will avoid left over the last cell in case
+  // that it must be clustered in one separate group
+  //how?
+  unsigned int k=1;
+  //firs_set, is a map, containing each compartment and their cell
+  // before the deagglomeration step
+  //since it is possible that some cells locate in same compartment but,
+  // they are not connected
+  //directly or indirectly and the need deagglomeration
```
[the remark this answers is not in the document]
Comment (voferreira) [reply source/rpt/compartmentalization.cc:155]: Agreed

File: include/rpt/compartmentalization.h [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+  void
+  test();
+
+private:
+  /**
+   * @brief Read the physical property (in this problem, the electric field)
+   * and write it into a vector of a vector.
+   */
+  auto
+  read_physical_property() -> std::vector<std::vector<double>>;
```
[the remark this answers is not in the document]
Comment (ghazalemir) [reply include/rpt/compartmentalization.h:63]: I was declaring it same way as you suggest, which works. But, this was the CLion suggestion and I approved its suggestion!

File: include/rpt/compartmentalization.h (hunk 1 above)
Comment (ghazalemir) [reply include/rpt/compartmentalization.h:66]: A map!

File: include/rpt/compartmentalization.h [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+
+  /**
+   * @brief Read the velocity magnitude and write it into a vector of a vector.
```
Comment (ghazalemir) [inline include/rpt/compartmentalization.h:66]:
```suggestion
   * @brief Read the velocity magnitude and write it into a map.
```

File: applications/compartmentalization_3d/compartmentalization_3d.cc [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+int
+main(int argc, char *argv[])
+{
+  try
+    {
+      Utilities::MPI::MPI_InitFinalize mpi_initialization(
+        argc, argv, numbers::invalid_unsigned_int);
+
+        Compartmentalization<3> compartmentalization;
+        compartmentalization.test();
```
[the remark this answers is not in the document]
Comment (ghazalemir) [reply applications/compartmentalization_3d/compartmentalization_3d.cc:34]: Done!

File: include/rpt/compartmentalization.h [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+#define lethe_Compartmentalization_h
+
+/**
+ * This class compartmentalize the a geometry/reactor based on physical
+ * values such as velocity, magnetic field,.... and calculate the
+ * flux between each compartment based on the velocity.
+ */
+
+#include <core/pvd_handler.h>
+#include <deal.II/base/config.h>
```
[the remark this answers is not in the document]
Comment (ghazalemir) [reply include/rpt/compartmentalization.h:27]: Done!

**Comment by blaisb** [comment]
@ghazalemir Do you want to keep this PR open? It's been open for too long now. You should try to find the time to finalize it.

**Pull request closed without merging**